Bottom: c5c269fa930d6b370358741973d3285f8eea309a
Top:    362b9773c155c4eae92102eb184e462167809f06
Author: Jason Wang <jasowang@redhat.com>
Date:   2017-03-02 14:50:30 +0800

Refresh of vhost-prefetch-desc-indices

---

diff --git a/drivers/vhost/vhost.c b/drivers/vhost/vhost.c
index 33003c7..cbe7f67 100644
--- a/drivers/vhost/vhost.c
+++ b/drivers/vhost/vhost.c
@@ -1910,11 +1910,12 @@ static int get_indirect(struct vhost_virtqueue *vq,
 
 /* Prefetch descriptor indices */
 int vhost_prefetch_desc_indices(struct vhost_virtqueue *vq,
-				__virtio16 *indices, int num)
+				__virtio16 *indices, u16 num)
 {
 	int ret;
-	u16 last_avail_idx, total, left;
+	u16 last_avail_idx, total, copied;
 	__virtio16 avail_idx;
+	__virtio16 *orig = indices;
 
 	if (unlikely(vhost_get_user(vq, avail_idx, &vq->avail->idx))) {
 		vq_err(vq, "Failed to access avail idx at %p\n",
@@ -1922,25 +1923,33 @@ int vhost_prefetch_desc_indices(struct vhost_virtqueue *vq,
 		return -EFAULT;
 	}
 	vq->avail_idx = vhost16_to_cpu(vq, avail_idx);
-	printk("num %d avail %d\n", num, vq->avail_idx - vq->last_avail_idx);
-	ret = total = min(num, vq->avail_idx - vq->last_avail_idx);
+	total = vq->avail_idx - vq->last_avail_idx;
+	printk("num %d avail %u\n", num, total);
+	ret = total = min(num, total);
 
 	last_avail_idx = vq->last_avail_idx & (vq->num - 1);
 	while (total) {
 		int ret2;
 
-		left = vq->num - last_avail_idx;
-		left = min(total, left);
+		copied = vq->num - last_avail_idx;
+		copied = min(total, copied);
 
+		if (indices + copied - orig > 64) {
+			printk("copied %d, before %d after %d\n",
+				copied, indices - orig,
+				indices + copied - orig);
+			return 0;
+		}
 		ret2 = vhost_copy_from_user(vq, indices,
 					    &vq->avail->ring[last_avail_idx],
-					    sizeof(avail_idx) * left);
+					    sizeof(avail_idx) * copied);
 		if (unlikely(ret2)) {
 			vq_err(vq, "Failed to get descriptors\n");
 			return -EFAULT;
 		}
 
-		total -= left;
+		indices += copied;
+		total -= copied;
 		last_avail_idx = 0;
 	}
 
diff --git a/drivers/vhost/vhost.h b/drivers/vhost/vhost.h
index c8a40af..c38acca 100644
--- a/drivers/vhost/vhost.h
+++ b/drivers/vhost/vhost.h
@@ -194,7 +194,7 @@ int vhost_get_vq_desc2(struct vhost_virtqueue *,
 void vhost_discard_vq_desc(struct vhost_virtqueue *, int n);
 
 int vhost_prefetch_desc_indices(struct vhost_virtqueue *vq,
-				__virtio16 *indices, int num);
+				__virtio16 *indices, u16 num);
 int vhost_vq_init_access(struct vhost_virtqueue *);
 int vhost_add_used(struct vhost_virtqueue *, unsigned int head, int len);
 int vhost_add_used_n(struct vhost_virtqueue *, struct vring_used_elem *heads,
